src/compute/src/sink/correction_v2.rs

```diff
@@ -0,0 +1,640 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::collections::VecDeque;
+use std::rc::Rc;
+
+use mz_persist_client::metrics::{SinkMetrics, SinkWorkerMetrics};
+use mz_repr::{Diff, Timestamp};
+use timely::container::columnation::{Columnation, TimelyStack};
+use timely::progress::Antichain;
+use timely::PartialOrder;
+
+const CHUNK_SIZE_BYTES: usize = 8 << 10;
+
+/// chains:
+///
+/// 0: xxxxxxxx
+/// 1: xxxx
+/// 2: xx
+/// 3: x
+///
+pub(super) struct Correction<D: Columnation> {
+    chains: Vec<Chain<D>>,
+    since: Antichain<Timestamp>,
+    /// Global persist sink metrics.
+    _metrics: SinkMetrics,
+    /// Per-worker persist sink metrics.
+    _worker_metrics: SinkWorkerMetrics,
+}
+
+impl<D: Clone + Ord + Columnation + 'static> Correction<D> {
+    /// Construct a new `Correction` instance.
+    pub fn new(metrics: SinkMetrics, worker_metrics: SinkWorkerMetrics) -> Self {
+        Self {
+            chains: Default::default(),
+            since: Antichain::from_elem(Timestamp::MIN),
+            _metrics: metrics,
+            _worker_metrics: worker_metrics,
+        }
+    }
+
+    /// Insert a batch of updates.
+    pub fn insert(&mut self, mut updates: Vec<(D, Timestamp, Diff)>) {
+        let Some(since_ts) = self.since.as_option() else {
+            // If the since is the empty frontier, discard all updates.
+            return;
+        };
+
+        for (_, time, _) in &mut updates {
+            *time = std::cmp::max(*time, *since_ts);
+        }
+
+        self.insert_inner(updates);
+    }
+
+    /// Insert a batch of updates, after negating their diffs.
+    pub fn insert_negated(&mut self, mut updates: Vec<(D, Timestamp, Diff)>) {
+        let Some(since_ts) = self.since.as_option() else {
+            // If the since is the empty frontier, discard all updates.
+            return;
+        };
+
+        for (_, time, diff) in &mut updates {
+            *time = std::cmp::max(*time, *since_ts);
+            *diff = -*diff;
+        }
+
+        self.insert_inner(updates);
+    }
+
+    fn insert_inner(&mut self, updates: Vec<(D, Timestamp, Diff)>) {
+        // TODO(optimization): append to the last chain if all times in `updates` are greater than
+        // all times in that chain
+
+        let chain = Chain::from(updates);
+        if !chain.is_empty() {
+            self.chains.push(chain);
+            self.restore_chain_invariant();
+        }
+    }
+
+    fn restore_chain_invariant(&mut self) {
+        let merge_needed = |chains: &[Chain<_>]| match chains {
+            [.., prev, last] => last.len() >= prev.len() / 2,
```

Comment:
You might want to use $/3$ here to avoid non-consolidating behavior. Dividing by two doesn't guarantee that the chains are compacted within any bound. (Chain 0 might be the negated sum of chains $1\ldots n$.)